x-pack/plugins/security/server/lib/authorization/disable_ui_capabilities.ts

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { mapValues } from 'lodash';
+import { UICapabilities } from 'ui/capabilities';
+import { Actions } from './actions';
+
+export function disableUICapabilitesFactory(server: any, request: any) {
+  const { authorization } = server.plugins.security;
+  const actions: Actions = authorization.actions;
+
+  return async function disableUICapabilities(uiCapabilities: UICapabilities) {
+    const uiActions = Object.entries(uiCapabilities).reduce<string[]>(
+      (acc, [featureId, featureUICapabilities]) => [
+        ...acc,
+        ...Object.keys(featureUICapabilities).map(uiCapability =>
+          actions.ui.get(featureId, uiCapability)
+        ),
+      ],
+      []
+    );
+
+    let checkPrivilegesResponse: any;
+    try {
+      const checkPrivilegesDynamically: any = authorization.checkPrivilegesDynamicallyWithRequest(
+        request
+      );
+      checkPrivilegesResponse = await checkPrivilegesDynamically(uiActions);
+    } catch (err) {
+      // if we get a 401/403, then we want to disable all uiCapabilities, as this
+      // is generally when the user hasn't authenticated yet and we're displaying the
+      // login screen, which isn't driven any uiCapabilities
+      if (err.statusCode === 401 || err.statusCode === 403) {
+        return mapValues(uiCapabilities, featureUICapabilities =>
+          mapValues(featureUICapabilities, () => false)
+        );
+      }
+      throw err;
+    }
+
+    return mapValues(uiCapabilities, (featureUICapabilities, featureId) => {
+      return mapValues(featureUICapabilities, (enabled, uiCapability) => {
+        // if the uiCapability has already been disabled, we don't want to re-enable it
+        if (!enabled) {
+          return false;
+        }
+
+        const action = actions.ui.get(featureId!, uiCapability!);
+        return checkPrivilegesResponse.privileges[action];
```

Comment:
Should we be handling the case where `checkPrivilegesResponse.privileges[action]` doesn't return a boolean? I think I'd feel safer with an explicit check. My followup question is: What should we do when we detect this scenario? I understand it _shouldn't_ happen, but it's my job to be paranoid 😄 

Maybe we just do something simple like `return checkPrivilegesResponse.privileges[action] === true`? If we needed to debug an issue, it might be helpful to have logging telling us that we got a bad response back. Or do we just fail to load the page in this scenario? So many options, and I'm just rambling at this point...

Reply:
We're doing validation here: https://github.com/elastic/kibana/blob/master/x-pack/plugins/security/server/lib/authorization/validate_es_response.js so we'd have to have messed something up in the checkPrivileges implementations themselves to make this the case, but checking it to `return checkPrivilegesResponse.privileges[action] === true` is definitely safer with almost no cost, I will make it so.

Reply:
That's right, I remember that validation now. Thanks!